### PR TITLE
 IO AP sensing (Follow up of request #219 and #218)

### PR DESCRIPTION
**Pull request by davechris**
- implemented generic .tra parser
- added a test
- added documentation to changelog
- solved the linting errors in apsensing.py sensornet.py and sensortran.py

And I run the tests:
- hatch run format --> successful
- hatch run test --> does still not work, but the command < pytest ./tests/ > worked OK, I guess. I do not know what to expect, the result was: 80 passed, 3 skipped, 1 xfailed, 1 xpassed, 22 warnings in 580.50s (0:09:40)
- hatch run docs:build --> build succeeded, 35 warnings  -- I guess this is OK?

### CODE REVIEW

**Comment by BSchilperoort** [comment]
Good to see you've found some time to work on this again! I've invited you as a collaborator, so the CI should run without me needing to click the button.

**Comment by davechris** [comment]
Yeah, nice to be back :) It took me some time, as I had a bunch of responsibilities last semester. It seems, that I get a lot of time for my research in the coming months.

Ah, that linter-config-change expains the complains with the printout commands.

File: src/dtscalibration/io/apsensing.py [hunk 1]
```diff
@@ -411,3 +423,263 @@ def metakey(meta, dict_to_parse, prefix):
...
+    tra_filelist : list of str
+
+    Returns:
+    --------
+    match_exists : boolean
+    file : str
+    """
+    match_exists = False
+    for _, file in enumerate(tra_filelist):
+        tra_timestamp = file[-18:-5]
```
Comment (BSchilperoort) [inline src/dtscalibration/io/apsensing.py:443]: A shame that they don't add the actual timestamp to the .tra file... Would make this more robust and less ugly. But I don't see a good way to avoid this.

Maybe one magical number less this way:
```suggestion
        tra_timestamp = file.replace(".tra","")[-13:]
```

Comment (BSchilperoort) [reply src/dtscalibration/io/apsensing.py:443]: Also, not looping through all files might be faster. This is functional, but for every .xml file you have to look at between 1 and _all_ .tra files.

```py
tra_files = "\n".join(tra_filelist) # make it one big string
tra_timestamps = re.findall(r"(\d{14}).tra", data)  # find 14 digits followed by .tra
if timestamp in tra_timestamps:
    return True
return False
```

The regex is explained here: https://regex101.com/r/B9al6Y/1

And then you can also remove the outer loop:
```py
tra_files = "\n".join(tra_filelist) # make it one big string
tra_timestamps = set(re.findall(r"(\d{14}).tra", data))  # find 14 digits followed by .tra

xml_timestamps = "\n".join(filepathlist)
xml_timestamps = set(re.findall(r"(\d{14}).xml", data)) # note that these are sets now

diff = xml_timestamps - tra_timestamps
if len(diff) > 0:
    msg = f"Not all .xml files have a matching .tra file.\n Missing are time following timestamps {diff}."
    raise ValueError(msg)

diff = tra_timestamps - xml_timestamps
if len(diff) > 0:
    msg = f"Not all .tra files have a matching .xml file.\n Missing are time following timestamps {diff}."
    raise ValueError(msg)
```

File: src/dtscalibration/io/apsensing.py [hunk 2]
```diff
@@ -411,3 +423,263 @@ def metakey(meta, dict_to_parse, prefix):
...
+
+    trace_key = [key for key in data_dict if "Trace." in key][
+        0
+    ]  # find key of trace n in "Trace.n" is unknown
+    data_dict["trace_key"] = trace_key
+
+    return data_dict
+
+
+def append_to_data_vars_structure(data_vars, data_dict_list):
```
Comment (BSchilperoort) [inline src/dtscalibration/io/apsensing.py:589]: This function doesn't feel very optimal. Lots of appending and modifying the numpy arrays... But I don't have time now to look into it deeper.

**Review by BSchilperoort — COMMENTED** [review]
I had a look, didn't run the code yet (but I assumes that part is fine).

The CI is complaining, and you probably don't see those errors because your hatch environments use less new versions of packages. If you do `hatch env prune` you remove them and hatch will create new environments that _should_ show the same errors as on the CI

File: src/dtscalibration/io/apsensing.py (hunk 2 above)
Comment (BSchilperoort) [reply src/dtscalibration/io/apsensing.py:589]: @davechris 

I rewrote this function. Should be faster and more legible now :)

```py
def append_to_data_vars_structure(data_vars, data_dict_list):
    """
    append data from .tra files to data_vars structure.
    (The data_vars structure is later on used to initialize the x-array dataset).


    Parameters
    ----------
    data_vars : dictionary containing *.xml data
    data_dict_list: list of dictionaries
                each dictionary in the list contains the data of one .tra file

    Returns:
    --------
    data_vars : dictionary containing *.xml data and *.tra data

    """
    # First derive numer of measurements in each file (should be constant):
    data_dict = data_dict_list[0]
    tr_key = data_dict["trace_key"]
    n_measurements = len([key for key in data_dict[tr_key] if isinstance(key, int)])
    
    # Initialize arrays
    distances = np.zeros((len(data_dict_list), n_measurements))
    t_by_dts = np.zeros((len(data_dict_list), n_measurements))
    log_ratio = np.zeros((len(data_dict_list), n_measurements))
    loss = np.zeros((len(data_dict_list), n_measurements))

    for idx, data_dict in enumerate(data_dict_list):
        # first get distance, t_by_dts, log_ratio and loss as list from dictionary
        tr_key = data_dict["trace_key"]
        data_keys = [key for key in data_dict[tr_key] if isinstance(key, int)]
        data = np.array([data_dict[tr_key][key] for key in data_keys]).T

        # Fill in pre-initialized array
        distances[idx] = data[0]
        t_by_dts[idx] = data[1]
        log_ratio[idx] = data[2]
        loss[idx] = data[3]

    # add log_ratio and attenaution to data_vars
    data_vars["log_ratio_by_dts"] = (("time", "x"), log_ratio)
    data_vars["loss_by_dts"] = (("time", "x"), loss)

    # add reference temp data, if they exist
    for idx_ref_temp in range(1, 5):
        if f"Ref.Temperature.Sensor.{idx_ref_temp}" in data_dict[tr_key]:
            ref_temps = []
            for _, data_dict in enumerate(data_dict_list):
                tr_key = data_dict["trace_key"]
                ref_temps.append(
                    data_dict[tr_key][f"Ref.Temperature.Sensor.{idx_ref_temp}"]
                )
            data_vars[f"probe{idx_ref_temp}Temperature"] = (("time",), ref_temps)

    # check if files match by comparing timestamps and dts temperature
    for idx_t in range(0, len(data_dict_list)):
        # check timestamps
        data_dict = data_dict_list[idx_t]
        tr_key = data_dict["trace_key"]
        dd_ts = pd.Timestamp(
            int(data_dict[tr_key]["Date.Year"]),
            int(data_dict[tr_key]["Date.Month"]),
            int(data_dict[tr_key]["Date.Day"]),
            int(data_dict[tr_key]["Time.Hour"]),
            int(data_dict[tr_key]["Time.Minute"]),
            int(data_dict[tr_key]["Time.Second"]),
        )

        err_msg = f"fatal error in allocation of .xml and .tra data.\nxml file {data_vars['creationDate'][1][idx_t]}\ntra file {str(dd_ts)}\n\n"
        if not data_vars["creationDate"][1][idx_t] == dd_ts:
            raise Exception(err_msg)

        # check dts temperature
        for idx_x in [0, 2, 5]:
            if not data_vars["tmp"][1][idx_x][idx_t] == t_by_dts[idx_t][idx_x]:
                # fatal error in allocation of .tra and .xml data
                raise Exception(err_msg)
    return data_vars

```

File: src/dtscalibration/io/apsensing.py [hunk 3]
```diff
@@ -83,7 +83,7 @@ def read_apsensing_files(
 
     device = apsensing_xml_version_check(filepathlist)
```
[the remark this answers is not in the document]
Comment (davechris) [reply src/dtscalibration/io/apsensing.py:84]: If users do not produce .xml files, they do not get stokes and anti-stokes intensity data and the use of this package is unfeasable.

File: src/dtscalibration/io/apsensing.py (hunk 2 above)
Comment (davechris) [reply src/dtscalibration/io/apsensing.py:589]: Thanks for rewriting the function!

Comment (davechris) [reply src/dtscalibration/io/apsensing.py:589]: I am questioning whether we need the trace data at all. 
- distances are not even transferred to data_vars
- t_by_dts is only used for a check (which is a bit unnecessary as we already check the time)
- log_ratio and loss are data we do not really need. Originally I was interested in comparing them do the self-calibrated data, but I never did it and feel that it is unnecessary, as I see already on the temperature curve, that the device calibration is not that good.

Comment (davechris) [reply src/dtscalibration/io/apsensing.py:589]: I will add a flag `load_full_tra = False` to read_apsensing_files. Thus, unnecessary trace data will not be imported in the default case.

File: tests/data/ap_sensing_2/readme_data_explanation.txt
[the remark this answers is not in the document]
Comment (davechris) [reply tests/data/ap_sensing_2/readme_data_explanation.txt:1]: Yes, done. :)

File: tests/test_datastore.py [hunk 4]
```diff
@@ -451,6 +452,26 @@ def test_read_apsensing_files():
     )
     np.testing.assert_almost_equal(ds.st.sum(), 10415.2837, decimal=2)
 
+    filepath_tra = data_dir_ap_sensing_with_tra
+    ds_tra = read_apsensing_files(
```
[the remark this answers is not in the document]
Comment (davechris) [reply tests/test_datastore.py:456]: Yes, done.

**Comment by davechris** [comment]
> The CI is complaining, and you probably don't see those errors because your hatch environments use less new versions of packages. If you do `hatch env prune` you remove them and hatch will create new environments that _should_ show the same errors as on the CI

 `hatch env prune` worked. So it seems that I can finally execute all hatch commands! 
 
 
I started with hatch run format:
_...\dtscalibration\lib\site-packages\bokeh\resources.py:363: error: Pattern matching is only supported in Python 3.10 and greater  [syntax]
Found 1 error in 1 file (errors prevented further checking)_

Either we need to change that pattern matching or require Python 3.10 in pyproject.toml:

- line 26 --> requires python 3.10 (from 3.9)
- line 131 delete "3.9" from array
- line 173 target version py310 statt py39

then it continues to search for errors, finds 10, fixes 8. The two remaining are identical:
UP038 in `assert isinstance(v, (list, tuple))` in src\dtscalibration\dts_accessor_utils.py
Very easy to fix (https://docs.astral.sh/ruff/rules/non-pep604-isinstance/)


The question is: Do you want to require python 3.10 or rather fix the first errror?

**Comment by BSchilperoort** [comment]
You can try changing the settings:
```toml
[tool.mypy]
ignore_missing_imports = true  # Preferably false, but matplotlib, scipy and statsmodels are missing typing stubs
python_version = "3.9"
exclude="*/site-packages/*"
```

**Comment by davechris** [comment]
Alright, I'll try that tomorrow.

**Review by BSchilperoort — APPROVED** [review]
CI is now green. If you can have a look at the delayed loading (into dask arrays) https://github.com/dtscalibration/python-dts-calibration/pull/220#discussion_r1743604855 and see if that still works or not (and add that info to docs), you can merge this PR.

File: src/dtscalibration/io/apsensing.py [hunk 5]
```diff
@@ -103,6 +103,16 @@ def read_apsensing_files(
...
+    tra_exists, tra_filepathlist = check_if_tra_exists(filepathlist)
+    if tra_exists:
+        print(".tra files exist and will be read")
+        data_dict_list = []
+        for _, tra_file in enumerate(tra_filepathlist):
+            data_dict = read_single_tra_file(tra_file)
+            data_dict_list.append(data_dict)
+        data_vars = append_to_data_vars_structure(data_vars, data_dict_list)
+
     ds = xr.Dataset(data_vars=data_vars, coords=coords, attrs=attrs, **kwargs)
```
[the remark this answers is not in the document]
Comment (davechris) [reply src/dtscalibration/io/apsensing.py]: - stated that in a sentence in the [docstring of io/apsensing.py](https://github.com/dtscalibration/python-dts-calibration/pull/220/commits/621333266d5af3730483a1bb25fd3ed20d3b7d05)
- stated that in a sentence and added a log file explanation and a .xml+.tra data import example to [docs/notebooks/A3Load_ap_sensing_files.ipynb](https://github.com/dtscalibration/python-dts-calibration/pull/220/commits/114e422aab6ac31041c4008fde8710927bddcb2d)

**Comment by davechris** [comment]
I added a remark to document that issue in the docstring and in the docs notebook A3.

And I do not have the rights to merge a PR ;)
